docs/devel/kubectl-conventions.md

```diff
@@ -135,6 +135,77 @@ Updated: 8/27/2015
 * Use "TYPE" for the particular flavor of resource type accepted by kubectl, rather than "RESOURCE" or "KIND"
 * Use "NAME" for resource names
 
+## Command structure conventions
+
+For every command there should be a `NewCmd<CommandName>` function that creates the command and returns a pointer to a `cobra.Command`, which can later be added to other parent commands to compose the structure tree. There should also be a `<CommandName>Options` struct with a variable to every flag and argument declared by the command (and any other variable required for the command to run). This makes tests and mocking easier. The struct ideally exposes three functions:
+
+* `Complete`: Completes the struct variables with values that may not be directly provided, for example, by flags pointers. Here you will usually take the `args` slice and set the values as appropriate variables, instantiate configs or clients, etc.
+* `Validate`: performs validation and returns errors.
+* `Run<CommandName>`: runs the actual command, taking as assumption that the struct is complete with all required values to run, and they are valid.
+
+Sample command skeleton:
+
+```go
+// MineOptions contains all the options for running the mine cli command
+type MineOptions struct {
+  mineLatest bool
+}
+
+const (
+  mineLong = `Some long description
+for my command.`
+
+  mineExample = `  # Run my command's first action
+  $ kubectl mine first
+
+  # Run my command's second action on latest stuff
+  $ kubectl mine second --latest`
+)
+
+// NewCmdMine implements the kubectl mine command.
+func NewCmdMine(f *cmdutil.Factory, out io.Writer) *cobra.Command {
```

Comment:
I agree with you though I have mixed feelings about pushing for this pattern here. Downstream we use two different parent names (`oc` and `openshift cli`) while here it's only `kubectl`. Is there any possibility in renaming/aliasing `kubectl` in the future? @bgrant0607 @brendandburns If not, I don't want to enforce an unnecessary convention.